internal/util/testutil/equal.go

```diff
@@ -318,3 +324,175 @@ func equalScalars(tb testing.TB, v1, v2 any) bool {
 		panic("not reached")
 	}
 }
+
+// equalValue compares any BSON values in a way that is useful for tests:
+//   - number types are converted so different types int32, int64, float64 are equal if the value is same.
+//
+// For more see equal.
+func equalValue(tb testing.TB, v1, v2 any) bool {
+	tb.Helper()
+
+	switch v1 := v1.(type) {
+	case *types.Document:
+		d, ok := v2.(*types.Document)
+		if !ok {
+			return false
+		}
+
+		return equalValueDocuments(tb, v1, d)
+
+	case *types.Array:
+		a, ok := v2.(*types.Array)
+		if !ok {
+			return false
+		}
+
+		return equalValueArrays(tb, v1, a)
+
+	default:
+		return equalValueScalars(tb, v1, v2)
+	}
+}
+
+// equalValueDocuments compares BSON documents and equates different number types.
+func equalValueDocuments(tb testing.TB, v1, v2 *types.Document) bool {
+	tb.Helper()
+
+	require.NotNil(tb, v1)
+	require.NotNil(tb, v2)
+
+	keys := v1.Keys()
+	if !slices.Equal(keys, v2.Keys()) {
+		return false
+	}
+
+	for _, k := range keys {
+		f1, err := v1.Get(k)
+		require.NoError(tb, err)
+
+		f2, err := v2.Get(k)
+		require.NoError(tb, err)
+
+		if !equalValue(tb, f1, f2) {
+			return false
+		}
+	}
+
+	return true
+}
+
+// equalValueArrays compares BSON arrays and equates different number types.
+func equalValueArrays(tb testing.TB, v1, v2 *types.Array) bool {
+	tb.Helper()
+
+	require.NotNil(tb, v1)
+	require.NotNil(tb, v2)
+
+	l := v1.Len()
+	if l != v2.Len() {
+		return false
+	}
+
+	for i := 0; i < l; i++ {
+		el1, err := v1.Get(i)
+		require.NoError(tb, err)
+
+		el2, err := v2.Get(i)
+		require.NoError(tb, err)
+
+		if !equalValue(tb, el1, el2) {
+			return false
+		}
+	}
+
+	return true
+}
+
+// equalValueScalars compares BSON scalar values and equates different number types.
+func equalValueScalars(tb testing.TB, v1, v2 any) bool {
```

Comment:
And this is a bit surprising to me, so even for scalars, `assert.EqualValues` is not enough?

Reply:
This part was removed.